### PR TITLE
Predefined messages in Bid Reject [reupload]

### PR DESCRIPTION
**Pull request by CryptoCubik**
Predefined messages specified via an enum on the command line for bid reject command.
Inserted into a "reason" bidDatas in the bid

### CODE REVIEW

File: src/api/commands/bid/BidRejectCommand.ts [hunk 1]
```diff
@@ -54,30 +56,45 @@ export class BidRejectCommand extends BaseCommand implements RpcCommandInterface
...
-     *
+     * [1]: reasonId
      * @param {RpcRequest} data
      * @returns {Promise<RpcRequest>}
      */
     public async validate(data: RpcRequest): Promise<RpcRequest> {
 
         if (data.params.length < 1) {
+            this.log.error('Missing bidId.');
             throw new MessageException('Missing bidId.');
```
Comment (xludx) [inline src/api/commands/bid/BidRejectCommand.ts]: fix the exceptions

File: src/api/commands/bid/BidRejectCommand.ts [hunk 2]
```diff
@@ -54,30 +56,45 @@ export class BidRejectCommand extends BaseCommand implements RpcCommandInterface
...
+            }
+        }
+
         const bidId = data.params[0];
-        const bid: resources.Bid = await this.bidService.findOne(bidId)
-            .then(value => {
-                return value.toJSON();
-            });
+        let bid: any = await this.bidService.findOne(bidId);
+        if (!bid) {
```
Comment (xludx) [inline src/api/commands/bid/BidRejectCommand.ts]: findone will throw NotFoundException unless something is found.

```
const bid: resources.Bid = await this.bidService.findOne(bidId)
            .then(value => value.toJSON());
```

File: src/api/enums/BidRejectReason.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+// Distributed under the GPL software license, see the accompanying
+// file COPYING or https://github.com/particl/particl-market/blob/develop/LICENSE
+
+/**
+ * BidRejectReason
+ *
+ */
+
+export enum BidRejectReason {
+    OUT_OF_STOCK = 'OUT_OF_STOCK'
```
Comment (xludx) [inline src/api/enums/BidRejectReason.ts]: should we add NO_REASON or something to be used as a default when no specific reason wasnt given?

File: src/api/services/BidActionService.ts [hunk 4]
```diff
@@ -735,14 +736,15 @@ export class BidActionService {
...
-
+    public async reject(bid: resources.Bid, reason: string): Promise<SmsgSendResponse> {
         if (bid.action === BidMessageType.MPA_BID) {
 
-            const listingItem = await this.listingItemService.findOne(bid.ListingItem.id, true)
-                .then(value => {
-                    return value.toJSON();
-                });
+            let listingItem: any = await this.listingItemService.findOne(bid.ListingItem.id, true);
+            if (!listingItem) {
```
Comment (xludx) [inline src/api/services/BidActionService.ts]: again unreachable code here..
```
const listingItem = await this.listingItemService.findOne(bid.ListingItem.id, true)
                .then(value => value.toJSON());
```

File: src/api/services/BidActionService.ts [hunk 5]
```diff
@@ -735,14 +736,15 @@ export class BidActionService {
...
+                this.log.error('Listing item not found.');
+                throw new MessageException('Listing item not found.');
```
Comment (xludx) [inline src/api/services/BidActionService.ts]: ModelNotFoundException

File: src/api/services/BidActionService.ts [hunk 6]
```diff
@@ -752,8 +754,14 @@ export class BidActionService {
             }
```
Comment (xludx) [inline src/api/services/BidActionService.ts]: unreachable code

File: src/api/services/BidActionService.ts [hunk 7]
```diff
@@ -752,8 +754,14 @@ export class BidActionService {
...
 
+            if (!BidRejectReason[reason]) {
+                this.log.error('Unknown BidRejectReason = ${reason}.');
+                throw new MessageException(`Unknown BidRejectReason = ${reason}.`);
+            }
+
             // create the bid reject message
-            const bidMessage = await this.bidFactory.getMessage(BidMessageType.MPA_REJECT, listingItem.hash);
+            const bidMessage = await this.bidFactory.getMessage(BidMessageType.MPA_REJECT, listingItem.hash,
+                [ { id: 'reason', value: reason } ] as IdValuePair[] );
```
Comment (xludx) [inline src/api/services/BidActionService.ts]: pass reason: string, not some IdValuePair[], its the factorys responsibility to hide the creation of that.

File: src/api/commands/bid/BidRejectCommand.ts [hunk 8]
```diff
@@ -54,30 +56,45 @@ export class BidRejectCommand extends BaseCommand implements RpcCommandInterface
...
+            throw new MessageException('Bid with specified bidId not found.');
+        }
+        bid = bid.toJSON();
 
         // make sure ListingItem exists
         if (_.isEmpty(bid.ListingItem)) {
-            this.log.error('ListingItem not found.');
-            throw new MessageException('ListingItem not found.');
+            this.log.error('ListingItem not found in bid.');
+            throw new MessageException('ListingItem not found in bid.');
```
Comment (xludx) [inline src/api/commands/bid/BidRejectCommand.ts]: use ModelNotFoundException

File: src/api/commands/bid/BidRejectCommand.ts [hunk 9]
```diff
@@ -54,30 +56,45 @@ export class BidRejectCommand extends BaseCommand implements RpcCommandInterface
...
         if (typeof data.params[0] !== 'number') {
+            this.log.error('bidId should be a number.');
             throw new MessageException('bidId should be a number.');
         }
 
+        if (data.params.length >= 2) {
+            const reason = data.params[1];
+            if (typeof reason !== 'string') {
+                this.log.error('reasonEnum should be a string.');
+                throw new MessageException('reasonEnum should be a string.');
```
Comment (xludx) [inline src/api/commands/bid/BidRejectCommand.ts]: InvalidParamException

File: src/api/commands/bid/BidRejectCommand.ts [hunk 10]
```diff
@@ -54,30 +56,45 @@ export class BidRejectCommand extends BaseCommand implements RpcCommandInterface
...
     public async validate(data: RpcRequest): Promise<RpcRequest> {
 
         if (data.params.length < 1) {
+            this.log.error('Missing bidId.');
             throw new MessageException('Missing bidId.');
         }
 
         if (typeof data.params[0] !== 'number') {
+            this.log.error('bidId should be a number.');
             throw new MessageException('bidId should be a number.');
```
Comment (xludx) [inline src/api/commands/bid/BidRejectCommand.ts]: InvalidParamException

**Comment by CryptoCubik** [comment]
All required changes should be made now

**Pull request merged**